File: app/src/main/java/com/example/client/api/ApiService.java
package com.example.client.api;

import com.example.client.model.ApiResponse;
import com.example.client.model.Product;

import java.util.List;

import okhttp3.MultipartBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.Path;

public interface ApiService {

    // product
    @GET("products")
    Call<ApiResponse<List<Product>>> getAllProducts();

    @GET("products/{id}")
    Call<ApiResponse<Product>> getProductById(@Path("id") String id);

    @POST("products")
    Call<ApiResponse<Product>> createProduct(@Body Product product);

    @PUT("products/{id}")
    Call<ApiResponse<Product>> updateProduct(@Path("id") String id, @Body Product product);

    @DELETE("products/{id}")
    Call<ApiResponse<Void>> deleteProduct(@Path("id") String id);

    @Multipart
    @PUT("products/{id}/image")
    Call<ApiResponse<Product>> uploadProductImage(@Path("id") String productId, @Part MultipartBody.Part imageFile);

}

File: app/src/main/java/com/example/socialblog/api/ApiService.kt
package com.example.socialblog.api

import com.example.socialblog.model.Post
import com.example.socialblog.model.User
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Path

interface ApiService {

    @POST("api/auth/register")
    fun register(
        @Body user: User
    ): Call<ApiResponse>

    @POST("api/auth/login")
    fun login(
        @Body user: User
    ): Call<ApiResponse>

    @GET("api/auth/user/{id}")
    fun getUser(
        @Path("id") id: Int
    ): Call<ApiResponse>

    @PUT("api/auth/update/{id}")
    fun updateUser(
        @Path("id") id: Int,
        @Body user: User
    ): Call<ApiResponse>

    @Multipart
    @POST("api/auth/upload/{id}")
    fun uploadAvatar(

        @Path("id") id: Int,

        @Part avatar: MultipartBody.Part

    ): Call<ApiResponse>

    @Multipart
    @POST("api/post/create")
    fun createPost(

        @Part("user_id")
        userId: RequestBody,

        @Part("content")
        content: RequestBody,

        @Part("feeling")
        feeling: RequestBody,

        @Part
        image: MultipartBody.Part?

    ): Call<ApiResponse>

    @GET("api/post/all/")
    fun getPosts(): Call<ApiResponse>
}